### DISCUSSION TITLE
Including types that have no fields cause validation errors

### DISCUSSION THREAD

**Discussion by mswiderski** [smallrye/smallrye-graphql#978] [Q&A]
Hi,

I am starting to work on building GraphQL for my services and run into a issue that I believe is caused by scaling all possible implementation of the interface and even adding the interface that can be just a marker interface - meaning no fields.

Here is a bit more context into the issue

- I have marker interface called `Model`
- this interface is implemented by `InvoiceInputModel` and `InvoiceOutputModel`
- these models are then used as a argument and return type of a GraphQL mutation 

So the first problem is that all of a sudden it fails with 

````
Caused by: graphql.AssertException: All types within a GraphQL schema must have unique names. No two provided types may have the same name.
No provided type may have a name which conflicts with any built in types (including Scalar and Introspection types).
You have redefined the type 'InvoiceInputModel' from being a 'GraphQLObjectType' to a 'GraphQLInputObjectType'
````

the problem seems to be caused by the fact that `InvoiceInputModel` is added twice - once when found as argument where it is added properly as input type to the schema, but then it is added once again as type based on scanning of the `Model` interface. 
Here I have first question - why is this actually happening? Shall it check if given type was already processed or at least check its annotation (as `InvoiceInputModel` is annotated with `Input`)?

I workaround it by removing model interface from the `InvoiceInputModel` and then it failed with another validation error

````
Caused by: graphql.schema.validation.InvalidSchemaException: invalid schema:
"Model" must define one or more fields.
````

Here I actually made a change in the SchemaBuilder to only add types that actually have fields or values (for enum). This allowed me to start the GraphQL service and it works nicely.

With that I'd like to ask about few things:
- the change I made in `SchemaBuilder` to skip types that do not have fields - is it ok or there should be different approach?
-  scanning of the interfaces of found type - shall it actually find all implementation of the interface as it might not be desired and could bring a lot of not needed stuff? Another example I ran into is I have JsonNode model that is in the code base but not used - comes with framework library. To make it work I had to add additional dependency indexation to not fail on it.

Any and all comments and hints are more needed :)

**Comment by phillip-kruger** [comment]
Typically when we want different input/output objects we use one model and annotate the relevant getters (for output) and setters (for input)/
So for example the `@Ignore` on a get field will omit the field on the output type but include it in the input type. A Input type will be created when the object is used as an input (parameter) and an output when returned from an operation.

Saying that, what you describe should also work, so maybe create a small reproducer I can look at, it's probably a bug

**Reply by mswiderski** [reply]
thanks for the feedback @phillip-kruger working on simple reproducer to show the issue with interfaces and classes implementing same interface for both input and output. Will post back when ready

**Comment by mswiderski** [comment]
here is a reproducer for the interface no fields issue https://github.com/mswiderski/smallrye-graphql-reproducer

let me know if there is anything else to do from my end.

**Reply by phillip-kruger** [reply]
Perfect ! Thanks, I'll have a look later this week and get back to you

**Comment by phillip-kruger** [comment]
@mswiderski - ok , I think we have multiple issues here. I'll reply here as I go 

For the empty Model interface that you have, we probably need to support the `@Ignore` on the Type (so we ignore the interface in GraphQL). The GraphQL Spec does not allow empty interfaces:

> An Interface type must define one or more fields

 see http://spec.graphql.org/draft/#sel-HAHbnBFBABABwBrqE

There is already a request to allow `@Ignore` on Type (see https://github.com/smallrye/smallrye-graphql/issues/503) but this is a MicroProfile Spec change, so might be harder to get done. Maybe I'll add a SmallRye Ignore for now until we can move this to the spec.

**Reply by mswiderski** [reply]
Could we (by default) ignore empty interfaces without the need to mark it with `@Ignore`? The reason for it is that some interfaces are out of control meaning they are imported from 3rd party libraries and thus can't be easily changed.

**Reply by phillip-kruger** [reply]
Good idea, let me look at that. You would not be able to use that interface anywhere in the operations

**Reply by phillip-kruger** [reply]
Ok, PR #989 will solve the empty interface. We will now ignore empty interfaces.

**Reply by mswiderski** [reply]
excellent, thanks

**Comment by phillip-kruger** [comment]
@mswiderski - Do you have a reproducer for the "added twice" issue ?

**Reply by mswiderski** [reply]
I was not able to reproduce it with the simple one but let me try to go back to the "full service" and try to spot the difference.

**Reply by mswiderski** [reply]
weird enough I cannot reproduce that any more so I think we can leave it. In case I would run into it again will provide more details and reproducer.

@phillip-kruger Thanks a lot for outstanding support on this!

**Reply by phillip-kruger** [reply]
No Problem ! Thanks for your help ! So I will release a new version of SmallRye GraphQL containing this fix to be included in Quarkus 2.3 or if we are lucky 2.2.1
